### ISSUE TITLE
auth API: Internal server error when deleting many RRSets in parallel with SQLite3 backend

### ISSUE THREAD

**Issue by jvns** [PowerDNS/pdns#14564]
- Program: Authoritative API
 - Issue type: Bug report

### Short description

When making many API requests all at the exact same time to delete records from the same zone, I get the following "database is locked" error:

```
HTTP ISE for "/api/v1/servers/localhost/zones/example.com.": 
Exception: GSQLBackend unable to delete RRSet test8.example.com|A:
Error while retrieving SQLite query results: database is locked (5)
```

I think what's happening is that multiple threads are trying to write to the database at the same time.

### Environment

 - Operating system: Mac OS
 - Software version: PowerDNS Authoritative Server 4.9.1
 - Software source: Homebrew

I'm seeing the same issue on Linux.

### Steps to reproduce

Here's a Python program and basic PowerDNS config that reproduces the issue: https://gist.github.com/jvns/cc23517f7c7d9156c2dbf649185caa96

Basically the Python program just creates 10 records and deletes them all at the same time.

### Expected behaviour

I'd expect it to synchronize the writes somehow, so that the request waits until the previous database write completed.

### Actual behaviour

There's an internal server error

**Comment by cmouse** [comment]
Can you try 
```
sqlite3 /path/to/db
PRAGMA journal_mode=WAL;
.quit
```

**Comment by zeha** [comment]
Just to be sure, could you add your config settings to this report?

**Comment by jvns** [comment]
I set the journal mode to WAL but got the same results.

**Comment by zeha** [comment]
Can easily confirm.

**Comment by zeha** [comment]
https://www.sqlite.org/wal.html#sometimes_queries_return_sqlite_busy_in_wal_mode

I think it might be smart in `GSQLBackend::startTransaction` to look at `domain_id` and if we expect writes to execute `BEGIN IMMEDIATE` instead of `BEGIN`. Will need a flag to SSQLite3::startTransaction.

(Somewhat similar to what LMBDBBackend does, I guess.)

**Comment by Habbie** [comment]
> I set the journal mode to WAL but got the same results.

with default settings, we already set the mode to WAL every startup

**Comment by Habbie** [comment]
just patching in a `BEGIN IMMEDIATE` indeed makes the problem go away.

**Comment by Habbie** [comment]
> I think it might be smart in `GSQLBackend::startTransaction` to look at `domain_id` and if we expect writes to execute `BEGIN IMMEDIATE` instead of `BEGIN`. Will need a flag to SSQLite3::startTransaction.

Having gone through all of our code, startTransaction almost always means we intend to write. `domain_id` does not appear to be a useful indicator - the conflicting writes in this test case, for example, all pass `-1`

**Comment by zeha** [comment]
> > I think it might be smart in `GSQLBackend::startTransaction` to look at `domain_id` and if we expect writes to execute `BEGIN IMMEDIATE` instead of `BEGIN`. Will need a flag to SSQLite3::startTransaction.
> 
> Having gone through all of our code, startTransaction almost always means we intend to write. `domain_id` does not appear to be a useful indicator - the conflicting writes in this test case, for example, all pass `-1`

After sleeping on my earlier comment, it occured to me that my suggested `domain_id` inspection logic was wrong. When startTransaction is called it can be assumed it will be a write tx.
